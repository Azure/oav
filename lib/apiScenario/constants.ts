import { ArmDeploymentScriptResource, ArmTemplate } from "./apiScenarioTypes";

export const DEFAULT_ARM_ENDPOINT = "https://management.azure.com";
export const DEFAULT_ARM_API_VERSION = "2020-06-01";
export const DEFAULT_ROLE_ASSIGNMENT_API_VERSION = "2022-04-01";

const armDeploymentScript: ArmDeploymentScriptResource = {
  type: "Microsoft.Resources/deploymentScripts",
  apiVersion: "2020-10-01",
  name: "",
  location: "[resourceGroup().location]",
  kind: "AzurePowerShell",
  identity: {
    type: "UserAssigned",
    userAssignedIdentities: {
      "[parameters('userAssignedIdentity')]": {},
    },
  },
  properties: {
    forceUpdateTag: "[parameters('utcValue')]",
    scriptContent: "",
    environmentVariables: [],
    timeout: "PT1H",
    cleanupPreference: "OnSuccess",
    retentionInterval: "P1D",
  },
};

export const armDeploymentScriptTemplate: ArmTemplate = {
  $schema: "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  contentVersion: "1.0.0.0",
  parameters: {
    utcValue: {
      type: "string",
      defaultValue: "[utcNow()]",
    },
    userAssignedIdentity: {
      type: "string",
      defaultValue: "$(userAssignedIdentity)",
    },
  },
  resources: [armDeploymentScript],
};
