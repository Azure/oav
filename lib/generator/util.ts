import * as fs from "fs";
import * as path from "path";
import { URL } from "url";

const ATTR_MAP: any = {
  location: "location",
  "azure-async-operation": "Azure-AsyncOperation",
  "original-uri": "original-uri",
};

export function getBaseName(filePath: string) {
  return path.basename(filePath);
}
export function isObject(definitionSpec: any) {
  return (
    definitionSpec.type === "object" || "properties" in definitionSpec || "allOf" in definitionSpec
  );
}
export function randomKey() {
  return `key${Math.ceil(Math.random() * 10000)}`;
}

export function getRp(specFilePath: string): string {
  return "microsoft." + specFilePath.split("/").slice(-1)[0].split(".")[0].toLowerCase();
}

export function fileExist(...pathName: string[]): boolean {
  return fs.existsSync(path.resolve(...pathName));
}

export function isLongRunning(specItem: any) {
  return specItem.content["x-ms-long-running-operation"];
}

export function readPayloadFile(payloadDir: string, operationId: string): any {
  if (!fileExist(payloadDir, operationId)) {
    return null;
  }
  const inputPath = path.resolve(payloadDir, operationId);
  const filenames: string[] = fs.readdirSync(inputPath).filter((filename: string) => {
    return (
      filename.endsWith(".json") &&
      ["200", "201", "202", "204"].includes(filename.split(".json")[0])
    );
  });
  if (filenames.length === 0) {
    return null;
  }
  const payload: any = {};
  filenames.forEach((filename: string) => {
    const statusCode = filename.split(".json")[0];
    const data = fs.readFileSync(path.resolve(inputPath, filename));
    payload[statusCode] = JSON.parse(data.toString());
  });
  return payload;
}

export function getPollingAttr(specItem: any) {
  if (!specItem.content["x-ms-long-running-operation"]) {
    return;
  }
  const lrOption = specItem.content["x-ms-long-running-operation-options"];
  if (lrOption && lrOption["final-state-via"]) {
    return ATTR_MAP[lrOption["final-state-via"]];
  }

  return "location";
}

export function getPollingUrl(payload: any, specItem: any) {
  let attr = "location";
  const asyncOperation: string = payload.liveResponse.headers["azure-AsyncOperation"];

  const lrOption = specItem.content["x-ms-long-running-operation-options"];
  let location = payload.liveResponse.headers.location;

  if (lrOption && lrOption["final-state-via"]) {
    attr = ATTR_MAP[lrOption["final-state-via"]];
    location = payload.liveResponse.headers[attr];
  } else if (asyncOperation) {
    location = asyncOperation;
  }
  const url = new URL(location);
  return url.pathname + url.search;
}

export function updateExmAndSpecFile(
  example: any,
  newSpec: any,
  specFilePath: string,
  exampleName: string
): any {
  const outputDir = path.resolve(specFilePath, "../", "examples");
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir);
  }
  const outputPath = path.resolve(outputDir, exampleName);
  console.log("example file path: " + outputPath);
  fs.writeFileSync(outputPath, JSON.stringify(example, null, 2), "utf8");
  if (newSpec) {
    //log.info("updated swagger file path: " + specFilePath);
    fs.writeFileSync(specFilePath, JSON.stringify(newSpec, null, 2), "utf8");
  }
}

export function referenceExmInSpec(
  specFilePath: string,
  apiPath: string,
  methodName: string,
  exampleName: string
): any {
  const data = fs.readFileSync(specFilePath);
  const spec = JSON.parse(data.toString());
  if (!spec.paths[apiPath][methodName]["x-ms-examples"]) {
    spec.paths[apiPath][methodName]["x-ms-examples"] = {};
  }
  if (!(exampleName in spec.paths[apiPath][methodName]["x-ms-examples"])) {
    spec.paths[apiPath][methodName]["x-ms-examples"][exampleName] = {
      $ref: `./examples/${exampleName}.json`,
    };
    return spec;
  }
}
