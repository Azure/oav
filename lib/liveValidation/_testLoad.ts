import "reflect-metadata";
import { LiveValidator } from "./liveValidator";

const main = async () => {
  const liveValidator = new LiveValidator(
    {
      directory: "/home/htc/azure-rest-api-specs",
      swaggerPathsPattern: ["specification/**/*.json"],
      loadValidatorInInitialize: true,
      loadValidatorInBackground: false,
      // useUnifiedModelCache: true,
      useJsonParser: false,
    },
    (msg, level, _meta) => {
      if (level !== "debug") {
        console.log(level, msg);
      }
    }
  );
  await liveValidator.initialize();
  global.gc();
  await new Promise((resolve) => setTimeout(resolve, 5000));
  debugger;
  global.gc();
  console.log(JSON.stringify(process.memoryUsage(), null, 2));
};

main().catch((err) => {
  console.error(err);
});
