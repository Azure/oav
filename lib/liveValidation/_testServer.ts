import "reflect-metadata";
import { createServer } from "http";
import { Readable } from "stream";
import { LiveValidator } from "./liveValidator";

const streamToString = (stream: Readable) => {
  return new Promise<string>((resolve, reject) => {
    let str = "";
    stream.on("data", (data) => {
      str += data;
    });
    stream.on("end", () => {
      resolve(str);
    });
    stream.on("error", (err) => {
      reject(err);
    });
  });
};

const main = async () => {
  const liveValidator = new LiveValidator(
    {
      directory: "/home/htc/azure-rest-api-specs",
      swaggerPathsPattern: ["specification/**/resource-manager/**/*.json"],
      loadValidatorInInitialize: true,
      loadValidatorInBackground: false,
      useUnifiedModelCache: true,
      useJsonParser: false,
    },
    (msg, level, _meta) => {
      if (level !== "debug") {
        console.log(level, msg);
      }
    }
  );

  const server = createServer(async (req, res) => {
    let result;

    try {
      const bodyStr = await streamToString(req);
      const body = JSON.parse(bodyStr);
      result = await liveValidator.validateLiveRequestResponse(body);
    } catch (e) {
      result = {
        message: e.message,
        stack: e.stack,
      };
    }
    const resultStr = JSON.stringify(result);

    res.writeHead(200, { "Content-Type": "text/json" });
    res.end(resultStr);
  });

  await liveValidator.initialize();
  server.listen(8080);
  await new Promise((resolve) => setTimeout(resolve, 500));
  console.log(JSON.stringify(process.memoryUsage(), null, 2));
};

main().catch((err) => {
  console.error(err);
});
