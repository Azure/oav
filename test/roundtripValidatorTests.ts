import * as assert from "assert";
import { DefaultConfig } from "../lib/util/constants";
import { RequestResponsePair } from "../lib/liveValidation/liveValidator";
import { LiveValidator } from "../lib/liveValidation/liveValidator";

jest.setTimeout(999999);

describe("Live Validator", () => {
  describe("Initialization", () => {
    it("OperationLoader should be completely initialized", async () => {
      console.log("OperationLoader should be completely initialized");
      const swaggerPattern = "specification/compute/resource-manager/Microsoft.Compute/stable/2021-11-01/runCommands.json";
      const glob = require("glob");
      const filePaths: string[] = glob.sync(swaggerPattern, {
        ignore: DefaultConfig.ExcludedExamplesAndCommonFiles,
        nodir: true,
      });
      const options = {
        directory: "./test/liveValidation/swaggers/specification",
        swaggerPathsPattern: [
          "compute/resource-manager/Microsoft.Compute/stable/2021-11-01/runCommands.json"
        ],
        swaggerPaths: filePaths,
        enableRoundTripValidator: true,
        excludedSwaggerPathsPattern: []
      };
      const validator = new LiveValidator(options);
      await validator.initialize();

      assert.equal(validator.withRefSibingsOperationSearcher.cache.size, 1);
    });

    it("OperationLoader should be completely initialized", async () => {
      console.log("OperationLoader should be completely initialized");
      const swaggerPattern = "specification/**/*.json";
      const glob = require("glob");
      const filePaths: string[] = glob.sync(swaggerPattern, {
        ignore: DefaultConfig.ExcludedExamplesAndCommonFiles,
        nodir: true,
      });
      const options = {
        directory: "./test/liveValidation/swaggers/specification",
        swaggerPathsPattern: [
          "**/*.json"
        ],
        swaggerPaths: filePaths,
        enableRoundTripValidator: true,
        excludedSwaggerPathsPattern: []
      };
      const validator = new LiveValidator(options);
      await validator.initialize();

      assert.equal(validator.withRefSibingsOperationSearcher.cache.size, 19);
    });

    it("readonly properties should not cause error", async () => {
      console.log("readonly properties should not cause error");
      const swaggerPattern = "specification/compute/resource-manager/Microsoft.Compute/stable/2021-11-01/*.json";
      const glob = require("glob");
      const filePaths: string[] = glob.sync(swaggerPattern, {
        ignore: DefaultConfig.ExcludedExamplesAndCommonFiles,
        nodir: true,
      });
      const options = {
        directory: "./test/liveValidation/swaggers/specification",
        swaggerPathsPattern: [
          "compute/resource-manager/Microsoft.Compute/stable/2021-11-01/*.json"
        ],
        swaggerPaths: filePaths,
        enableRoundTripValidator: true,
        excludedSwaggerPathsPattern: []
      };
      const validator = new LiveValidator(options);
      await validator.initialize();

      //roundtrip validation
      const payload: RequestResponsePair = require(`${__dirname}/liveValidation/payloads/roundTrip_valid.json`);
      const rest = await validator.validateRoundTrip(payload);
      assert.equal(rest.errors, 0);
      assert.equal(rest.isSuccessful, true);
      expect(rest).toMatchSnapshot();
      //end of roundtrip validation
    });

    it("Round trip validation fail", async () => {
      console.log("Round trip validation fail");
      const swaggerPattern = "specification/compute/resource-manager/Microsoft.Compute/stable/2021-11-01/*.json";
      const glob = require("glob");
      const filePaths: string[] = glob.sync(swaggerPattern, {
        ignore: DefaultConfig.ExcludedExamplesAndCommonFiles,
        nodir: true,
      });
      const options = {
        directory: "./test/liveValidation/swaggers/specification",
        swaggerPathsPattern: [
          "compute/resource-manager/Microsoft.Compute/stable/2021-11-01/*.json"
        ],
        swaggerPaths: filePaths,
        enableRoundTripValidator: true,
        excludedSwaggerPathsPattern: []
      };
      const validator = new LiveValidator(options);
      await validator.initialize();

      //roundtrip validation
      const payload: RequestResponsePair = require(`${__dirname}/liveValidation/payloads/roundTrip_invalid.json`);
      const rest = await validator.validateRoundTrip(payload);
      expect(rest).toMatchSnapshot();
      assert.equal(rest.errors.length, 7);
      assert.equal(rest.isSuccessful, false);
      for (const re of rest.errors) {
        if (re.pathsInPayload[0].includes("location")) {
          assert.equal(re.code, "ROUNDTRIP_INCONSISTENT_PROPERTY");
        } else if (re.pathsInPayload[0].includes("createOption")) {
          assert.equal(re.code, "ROUNDTRIP_MISSING_PROPERTY");
        } else if (re.pathsInPayload[0].includes("caching")) {
          assert.equal(re.code, "ROUNDTRIP_ADDITIONAL_PROPERTY");
        } else if (re.pathsInPayload[0].includes("offer")) {
          assert.equal(re.code, "ROUNDTRIP_MISSING_PROPERTY");
        } else if (re.pathsInPayload[0].includes("computerName")) {
          assert.equal(re.code, "ROUNDTRIP_MISSING_PROPERTY");
        }
      }
      //end of roundtrip validation
    });

    it("Round trip validation of circular spec", async () => {
      console.log("Round trip validation fail");
      const swaggerPattern = "specification/containerservice/resource-manager/Microsoft.ContainerService/stable/2019-08-01/*.json";
      const glob = require("glob");
      const filePaths: string[] = glob.sync(swaggerPattern, {
        ignore: DefaultConfig.ExcludedExamplesAndCommonFiles,
        nodir: true,
      });
      const options = {
        directory: "./test/liveValidation/swaggers/specification",
        swaggerPathsPattern: [
          "containerservice/resource-manager/Microsoft.ContainerService/stable/2019-08-01/*.json"
        ],
        swaggerPaths: filePaths,
        enableRoundTripValidator: true,
        excludedSwaggerPathsPattern: []
      };
      const validator = new LiveValidator(options);
      await validator.initialize();

      //roundtrip validation
      const payload: RequestResponsePair = require(`${__dirname}/liveValidation/payloads/roundtrip_failure_circularspec.json`);
      const rest = await validator.validateRoundTrip(payload);
      assert.equal(rest.errors.length, 3);
      assert.equal(rest.isSuccessful, false);
      for (const re of rest.errors) {
        if (re.pathsInPayload[0].includes("location")) {
          assert.equal(re.code, "ROUNDTRIP_ADDITIONAL_PROPERTY");
        } else if (re.pathsInPayload[0].includes("properties")) {
          assert.equal(re.code, "ROUNDTRIP_ADDITIONAL_PROPERTY");
        } else if (re.pathsInPayload[0].includes("identity")) {
          assert.equal(re.code, "ROUNDTRIP_ADDITIONAL_PROPERTY");
        }
      }
      //end of roundtrip validation
    });

    it("Round trip validation of circular spec cognitiveService", async () => {
      console.log("Round trip validation fail");
      const swaggerPattern = "specification/cognitiveservices/data-plane/Language/preview/2022-10-01-preview/*.json";
      const glob = require("glob");
      const filePaths: string[] = glob.sync(swaggerPattern, {
        ignore: DefaultConfig.ExcludedExamplesAndCommonFiles,
        nodir: true,
      });
      const options = {
        directory: "./test/liveValidation/swaggers/specification",
        swaggerPathsPattern: [
          "cognitiveservices/data-plane/Language/preview/2022-10-01-preview/*.json"
        ],
        swaggerPaths: filePaths,
        enableRoundTripValidator: true,
        excludedSwaggerPathsPattern: []
      };
      const validator = new LiveValidator(options);
      await validator.initialize();

      //roundtrip validation
      const payload: RequestResponsePair = require(`${__dirname}/liveValidation/payloads/roundTrip_circularReference_analyzetext.json`);
      const rest = await validator.validateRoundTrip(payload);
      assert.equal(rest.errors.length, 5);
      assert.equal(rest.isSuccessful, false);
      for (const re of rest.errors) {
        if (re.pathsInPayload[0].includes("text")) {
          assert.equal(re.code, "ROUNDTRIP_ADDITIONAL_PROPERTY");
        } else if (re.pathsInPayload[0].includes("defaultLanguage")) {
          assert.equal(re.code, "ROUNDTRIP_INCONSISTENT_PROPERTY");
        } else if (re.pathsInPayload[0].includes("version")) {
          assert.equal(re.code, "ROUNDTRIP_MISSING_PROPERTY");
        }
      }
      //end of roundtrip validation
    });

  });
});
