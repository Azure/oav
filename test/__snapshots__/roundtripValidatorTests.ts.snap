// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Live Validator Initialization OperationLoader should be completely initialized 1`] = `
Array [
  "location",
  "/200/schema/properties/nextLink",
  "/200/schema/properties/value/items/properties/osType",
]
`;

exports[`Live Validator Initialization OperationLoader should be completely initialized for outer circular ref 1`] = `
Object {
  "operationId": "AnalyzeText",
  "parameters": Array [
    Object {
      "in": "query",
      "name": "api-version",
      "required": true,
      "type": "string",
    },
    Object {
      "in": "query",
      "name": "showStats",
      "required": false,
      "type": "boolean",
      "x-ms-parameter-location": "method",
    },
    Object {
      "in": "body",
      "name": "body",
      "required": true,
      "schema": Object {
        "discriminator": "kind",
        "properties": Object {
          "kind": Object {
            "enum": Array [
              "SentimentAnalysis",
              "EntityRecognition",
              "PiiEntityRecognition",
              "KeyPhraseExtraction",
              "LanguageDetection",
              "EntityLinking",
              "DynamicClassification",
            ],
            "type": "string",
            "x-ms-enum": Object {
              "modelAsString": true,
              "name": "AnalyzeTextTaskKind",
              Symbol(./source-map/object-info): [Function],
            },
            Symbol(./source-map/object-info): [Function],
          },
          Symbol(./source-map/object-info): [Function],
        },
        "required": Array [
          "kind",
        ],
        "type": "object",
        Symbol(./source-map/object-info): [Function],
      },
      Symbol(./source-map/object-info): [Function],
    },
  ],
  "responses": Object {
    "200": Object {
      "schema": Object {
        "discriminator": "kind",
        "properties": Object {
          "kind": Object {
            "enum": Array [
              "SentimentAnalysisResults",
              "EntityRecognitionResults",
              "PiiEntityRecognitionResults",
              "KeyPhraseExtractionResults",
              "LanguageDetectionResults",
              "EntityLinkingResults",
              "DynamicClassificationResults",
            ],
            "type": "string",
            "x-ms-enum": Object {
              "modelAsString": true,
              "name": "AnalyzeTextTaskResultsKind",
              Symbol(./source-map/object-info): [Function],
            },
            Symbol(./source-map/object-info): [Function],
          },
          Symbol(./source-map/object-info): [Function],
        },
        "required": Array [
          "kind",
        ],
        "type": "object",
        Symbol(./source-map/object-info): [Function],
      },
      Symbol(./source-map/object-info): [Function],
    },
    "default": Object {
      "schema": Object {
        "additionalProperties": false,
        "properties": Object {
          "error": Object {
            "$ref": "#/definitions/AnalyzeTextJobState/allOf/0/properties/errors/items",
          },
        },
        "required": Array [
          "error",
        ],
        "type": "object",
      },
      Symbol(./source-map/object-info): [Function],
    },
    Symbol(./source-map/object-info): [Function],
  },
}
`;

exports[`Live Validator Initialization OperationLoader should be initialized only with spec 1`] = `
Array [
  "parameters/schema/properties/properties/properties/statuses",
  "parameters/schema/allOf/0/properties/id",
  "parameters/schema/allOf/0/properties/name",
  "parameters/schema/allOf/0/properties/type",
  "/200/schema/properties/properties/properties/statuses",
  "/200/schema/allOf/0/properties/id",
  "/200/schema/allOf/0/properties/name",
  "/200/schema/allOf/0/properties/type",
]
`;

exports[`Live Validator Initialization OperationLoader should be initialized only with spec 2`] = `
Map {
  "readOnly" => Map {
    "parameters/schema/properties/properties/properties/statuses" => Array [],
    "parameters/schema/allOf/0/properties/id" => Array [],
    "parameters/schema/allOf/0/properties/name" => Array [],
    "parameters/schema/allOf/0/properties/type" => Array [],
    "/200/schema/properties/properties/properties/statuses" => Array [],
    "/200/schema/allOf/0/properties/id" => Array [],
    "/200/schema/allOf/0/properties/name" => Array [],
    "/200/schema/allOf/0/properties/type" => Array [],
  },
  "secret" => Map {},
  "default" => Map {},
  "mutability" => Map {
    "parameters/schema/properties/sku" => Array [
      "read",
    ],
    "/200/schema/properties/sku" => Array [
      "read",
    ],
  },
}
`;

exports[`Live Validator Initialization Round trip validation fail 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ROUNDTRIP_INCONSISTENT_PROPERTY",
      "jsonPathsInPayload": Array [],
      "message": "The property's value 'westcentralus1' in the GET response is different from what was set 'westcentralus' in the preceding PUT request.",
      "pathsInPayload": Array [
        "/location",
      ],
      "schemaPath": "",
      "severity": 0,
      "source": Object {
        "position": Object {
          "column": 0,
          "line": 0,
        },
        "url": "",
      },
    },
    Object {
      "code": "ROUNDTRIP_ADDITIONAL_PROPERTY",
      "jsonPathsInPayload": Array [],
      "message": "The property 'versionlist/1' is returned in the GET response, but it is not declared in the PUT request.",
      "pathsInPayload": Array [
        "/versionlist/1",
      ],
      "schemaPath": "",
      "severity": 0,
      "source": Object {
        "position": Object {
          "column": 0,
          "line": 0,
        },
        "url": "",
      },
    },
    Object {
      "code": "ROUNDTRIP_MISSING_PROPERTY",
      "jsonPathsInPayload": Array [],
      "message": "The property 'createOption' is present in the PUT request but is either never returned in the GET response or is returned with a null value.",
      "pathsInPayload": Array [
        "/properties/storageProfile/osDisk/createOption",
      ],
      "schemaPath": "",
      "severity": 0,
      "source": Object {
        "position": Object {
          "column": 0,
          "line": 0,
        },
        "url": "",
      },
    },
    Object {
      "code": "ROUNDTRIP_ADDITIONAL_PROPERTY",
      "jsonPathsInPayload": Array [],
      "message": "The property 'caching' is returned in the GET response, but it is not declared in the PUT request.",
      "pathsInPayload": Array [
        "/properties/storageProfile/osDisk/caching",
      ],
      "schemaPath": "",
      "severity": 0,
      "source": Object {
        "position": Object {
          "column": 0,
          "line": 0,
        },
        "url": "",
      },
    },
    Object {
      "code": "ROUNDTRIP_MISSING_PROPERTY",
      "jsonPathsInPayload": Array [],
      "message": "The property 'offer' is present in the PUT request but is either never returned in the GET response or is returned with a null value.",
      "pathsInPayload": Array [
        "/properties/storageProfile/imageReference/offer",
      ],
      "schemaPath": "",
      "severity": 0,
      "source": Object {
        "position": Object {
          "column": 0,
          "line": 0,
        },
        "url": "",
      },
    },
    Object {
      "code": "ROUNDTRIP_MISSING_PROPERTY",
      "jsonPathsInPayload": Array [],
      "message": "The property 'computerName' is present in the PUT request but is either never returned in the GET response or is returned with a null value.",
      "pathsInPayload": Array [
        "/properties/osProfile/computerName",
      ],
      "schemaPath": "",
      "severity": 0,
      "source": Object {
        "position": Object {
          "column": 0,
          "line": 0,
        },
        "url": "",
      },
    },
    Object {
      "code": "ROUNDTRIP_ADDITIONAL_PROPERTY",
      "jsonPathsInPayload": Array [],
      "message": "The property 'hardwareProfile' is returned in the GET response, but it is not declared in the PUT request.",
      "pathsInPayload": Array [
        "/hardwareProfile",
      ],
      "schemaPath": "",
      "severity": 0,
      "source": Object {
        "position": Object {
          "column": 0,
          "line": 0,
        },
        "url": "",
      },
    },
  ],
  "isSuccessful": false,
  "operationInfo": Object {
    "apiVersion": "2021-11-01",
    "operationId": "VirtualMachines_CreateOrUpdate",
  },
  "runtimeException": undefined,
}
`;

exports[`Live Validator Initialization Round trip validation of circular spec 1`] = `
Array [
  "/200/schema/allOf/0/properties/id",
  "/200/schema/allOf/0/properties/name",
  "/200/schema/allOf/0/properties/type",
  "/200/schema/allOf/1/properties/properties/properties/provisioningState",
  "/200/schema/allOf/1/properties/properties/properties/maxAgentPools",
  "/200/schema/allOf/1/properties/properties/properties/fqdn",
  "/200/schema/allOf/1/properties/properties/properties/agentPoolProfiles/items/allOf/0/properties/provisioningState",
  "/200/schema/allOf/1/properties/identity/properties/principalId",
  "/200/schema/allOf/1/properties/identity/properties/tenantId",
]
`;

exports[`Live Validator Initialization Round trip validation of circular spec 2`] = `
Object {
  "requestValidationResult": Object {
    "errors": Array [],
    "isSuccessful": true,
    "operationInfo": Object {
      "apiVersion": "2019-08-01",
      "operationId": "ManagedClusters_Get",
    },
    "runtimeException": undefined,
  },
  "responseValidationResult": Object {
    "errors": Array [
      Object {
        "code": "INVALID_CONTENT_TYPE",
        "documentationUrl": "",
        "jsonPathsInPayload": Array [],
        "message": "Invalid Content-Type (application/octet-stream).  These are supported: application/json",
        "pathsInPayload": Array [],
        "schemaPath": "",
        "severity": 1,
        "source": Object {
          "position": Object {
            "column": -1,
            "line": -1,
          },
          "url": "",
        },
      },
      Object {
        "code": "OBJECT_MISSING_REQUIRED_PROPERTY",
        "documentationUrl": "",
        "jsonPathsInPayload": Array [
          "$.properties.linuxProfile.ssh",
        ],
        "message": "Missing required property: ssh",
        "params": Array [
          "ssh",
        ],
        "pathsInPayload": Array [
          "/properties/linuxProfile/ssh",
        ],
        "schemaPath": "#/required",
        "severity": 0,
        "source": Object {
          "jsonRef": "#/definitions/ContainerServiceLinuxProfile",
          "position": Object {
            "column": 37,
            "line": 1547,
          },
          "url": "containerservice/resource-manager/Microsoft.ContainerService/stable/2019-08-01/managedClusters.json",
        },
      },
      Object {
        "code": "PATTERN",
        "documentationUrl": "",
        "jsonPathsInPayload": Array [
          "$.properties.networkProfile.podCidr",
        ],
        "message": "String does not match pattern ^([0-9]{1,3}\\\\.){3}[0-9]{1,3}(\\\\/([0-9]|[1-2][0-9]|3[0-2]))?$: aaaaaaaaaaaaa",
        "params": Array [
          "^([0-9]{1,3}\\\\.){3}[0-9]{1,3}(\\\\/([0-9]|[1-2][0-9]|3[0-2]))?$",
          "aaaaaaaaaaaaa",
        ],
        "pathsInPayload": Array [
          "/properties/networkProfile/podCidr",
        ],
        "schemaPath": "#/properties/podCidr/pattern",
        "severity": 0,
        "source": Object {
          "jsonRef": "#/definitions/ContainerServiceNetworkProfile/properties/podCidr",
          "position": Object {
            "column": 20,
            "line": 1592,
          },
          "url": "containerservice/resource-manager/Microsoft.ContainerService/stable/2019-08-01/managedClusters.json",
        },
      },
      Object {
        "code": "PATTERN",
        "documentationUrl": "",
        "jsonPathsInPayload": Array [
          "$.properties.networkProfile.serviceCidr",
        ],
        "message": "String does not match pattern ^([0-9]{1,3}\\\\.){3}[0-9]{1,3}(\\\\/([0-9]|[1-2][0-9]|3[0-2]))?$: aaaaaaaaaaa",
        "params": Array [
          "^([0-9]{1,3}\\\\.){3}[0-9]{1,3}(\\\\/([0-9]|[1-2][0-9]|3[0-2]))?$",
          "aaaaaaaaaaa",
        ],
        "pathsInPayload": Array [
          "/properties/networkProfile/serviceCidr",
        ],
        "schemaPath": "#/properties/serviceCidr/pattern",
        "severity": 0,
        "source": Object {
          "jsonRef": "#/definitions/ContainerServiceNetworkProfile/properties/serviceCidr",
          "position": Object {
            "column": 24,
            "line": 1598,
          },
          "url": "containerservice/resource-manager/Microsoft.ContainerService/stable/2019-08-01/managedClusters.json",
        },
      },
      Object {
        "code": "PATTERN",
        "documentationUrl": "",
        "jsonPathsInPayload": Array [
          "$.properties.networkProfile.dnsServiceIP",
        ],
        "message": "String does not match pattern ^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$: aaaaaaaaa",
        "params": Array [
          "^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$",
          "aaaaaaaaa",
        ],
        "pathsInPayload": Array [
          "/properties/networkProfile/dnsServiceIP",
        ],
        "schemaPath": "#/properties/dnsServiceIP/pattern",
        "severity": 0,
        "source": Object {
          "jsonRef": "#/definitions/ContainerServiceNetworkProfile/properties/dnsServiceIP",
          "position": Object {
            "column": 25,
            "line": 1604,
          },
          "url": "containerservice/resource-manager/Microsoft.ContainerService/stable/2019-08-01/managedClusters.json",
        },
      },
      Object {
        "code": "PATTERN",
        "documentationUrl": "",
        "jsonPathsInPayload": Array [
          "$.properties.networkProfile.dockerBridgeCidr",
        ],
        "message": "String does not match pattern ^([0-9]{1,3}\\\\.){3}[0-9]{1,3}(\\\\/([0-9]|[1-2][0-9]|3[0-2]))?$: aaaaaaaaaaaaa",
        "params": Array [
          "^([0-9]{1,3}\\\\.){3}[0-9]{1,3}(\\\\/([0-9]|[1-2][0-9]|3[0-2]))?$",
          "aaaaaaaaaaaaa",
        ],
        "pathsInPayload": Array [
          "/properties/networkProfile/dockerBridgeCidr",
        ],
        "schemaPath": "#/properties/dockerBridgeCidr/pattern",
        "severity": 0,
        "source": Object {
          "jsonRef": "#/definitions/ContainerServiceNetworkProfile/properties/dockerBridgeCidr",
          "position": Object {
            "column": 29,
            "line": 1610,
          },
          "url": "containerservice/resource-manager/Microsoft.ContainerService/stable/2019-08-01/managedClusters.json",
        },
      },
      Object {
        "code": "ENUM_CASE_MISMATCH",
        "documentationUrl": "",
        "jsonPathsInPayload": Array [
          "$.properties.networkProfile.loadBalancerSku",
        ],
        "message": "Enum does not match case for: Standard",
        "params": Array [
          "Standard",
          Array [
            "standard",
            "basic",
          ],
        ],
        "pathsInPayload": Array [
          "/properties/networkProfile/loadBalancerSku",
        ],
        "schemaPath": "#/properties/loadBalancerSku/enum",
        "severity": 0,
        "source": Object {
          "jsonRef": "#/definitions/ContainerServiceNetworkProfile/properties/loadBalancerSku",
          "position": Object {
            "column": 28,
            "line": 1616,
          },
          "url": "containerservice/resource-manager/Microsoft.ContainerService/stable/2019-08-01/managedClusters.json",
        },
      },
      Object {
        "code": "ENUM_MISMATCH",
        "documentationUrl": "",
        "jsonPathsInPayload": Array [
          "$.identity.type",
        ],
        "message": "No enum match for: UserAssigned",
        "params": Array [
          "UserAssigned",
          Array [
            "SystemAssigned",
            "None",
          ],
        ],
        "pathsInPayload": Array [
          "/identity/type",
        ],
        "schemaPath": "#/properties/type/enum",
        "severity": 0,
        "source": Object {
          "jsonRef": "#/definitions/ManagedClusterIdentity/properties/type",
          "position": Object {
            "column": 17,
            "line": 1893,
          },
          "url": "containerservice/resource-manager/Microsoft.ContainerService/stable/2019-08-01/managedClusters.json",
        },
      },
    ],
    "isSuccessful": false,
    "operationInfo": Object {
      "apiVersion": "2019-08-01",
      "operationId": "ManagedClusters_Get",
    },
    "runtimeException": undefined,
  },
}
`;

exports[`Live Validator Initialization readonly properties should not cause error 1`] = `
Object {
  "errors": Array [],
  "isSuccessful": true,
  "operationInfo": Object {
    "apiVersion": "2021-11-01",
    "operationId": "VirtualMachines_CreateOrUpdate",
  },
  "runtimeException": undefined,
}
`;
